net/Socket.hpp

```diff
@@ -493,6 +512,53 @@ class Socket
 
     /// We check the owner even in the release builds, needs to be always correct.
     std::thread::id _owner;
+
+    /// true if counted against net::Defaults.maxTCPConnections and dtor must call `decrTCPConnCount`
+    /// post socket ::close where the socket has been returned to the system.
+    bool _isCounted;
+
+    /// Decrements global connection count
+    /// Call this after socket ::close, where it has been returned to the system.
+    /// @param fd the related file descriptor for logging
+    static void decrTCPConnCount(int fd)
+    {
+        std::lock_guard<std::mutex> lock(maxTCPConnMutex);
+        const size_t u = maxTCPConnCount;
+        auto logPrefix = [fd](std::ostream& os) { os << '#' << fd << ": "; };
+        if (u > 0)
+        {
+            const size_t v = --maxTCPConnCount;
+            LOG_TRC("TCP Limiter: Count decremented: " << v);
+        }
+        else
+            LOG_WRN("TCP Limiter: Count decrement underflow: " << u);
+    }
+
+    /// Increments global TCP connection counter if not exceeding net::DefaultValue::maxTCPConnections.
+    /// Returns true if free connections were available, otherwise false.
+    /// No limitation is applied if net::DefaultValue::maxTCPConnections == 0.
+    /// @param fd the related file descriptor for logging
+    static bool checkAndIncrTCPConnCount(int fd)
+    {
+        std::lock_guard<std::mutex> lock(maxTCPConnMutex);
+        const size_t u = maxTCPConnCount;
+        auto logPrefix = [fd](std::ostream& os) { os << '#' << fd << ": "; };
+        if (net::Defaults.maxTCPConnections == 0 || u < net::Defaults.maxTCPConnections)
+        {
+            const size_t v = ++maxTCPConnCount;
+            LOG_TRC("TCP Limiter: Count incremented: " << v);
+            return true;
+        }
+        else
+        {
+            LOG_WRN("TCP Limiter: Limit reached: " << u);
+            return false;
+        }
+    }
+    friend class StreamSocket; // allow `checkAndIncrConnectionCount`
+
+    static std::mutex maxTCPConnMutex;
+    static size_t maxTCPConnCount; // accepted TCP IPv4/IPv6 socket count
```

Comment:
Lets put this in StreamSocket as a static std::atomic, and call it ExternalSocketCount or whatever we want to limit based on - if there is some consistent naming for it - based on whatever we found when looking at the relevant naming of Apache / NGINX settings.

Reply:
re StreamSocket ctor: See comments 'dtor close' above and 'TSocket leak' below. 
If it is deemed acceptable .. we can move this to the ctor, otherwise better stays.

If there is no such TSocket using the FD .. it would be OK'sh .. I guess.

Reply:
> If there is no such TSocket using the FD
I expect you to read all the sub-classes, and take a look - as part of researching the most elegant and minimal solution - it should be possible to answer that in 5-10 minutes.